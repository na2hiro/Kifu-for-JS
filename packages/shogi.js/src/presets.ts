import Color from "./Color";

const EMPTY = " *  *  *  *  *  *  *  *  * ";
const BOARD3_9 = [
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    EMPTY,
    EMPTY,
    EMPTY,
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
];
const BOARD2_9 = [EMPTY].concat(BOARD3_9);

/**
 * 既定の初期局面
 */
const presetDefinitions = {
    HIRATE: {
        board: ["-KY-KE-GI-KI-OU-KI-GI-KE-KY", " * -HI *  *  *  *  * -KA * "].concat(BOARD3_9),
        turn: Color.Black,
    },
    KY: {
        board: ["-KY-KE-GI-KI-OU-KI-GI-KE * ", " * -HI *  *  *  *  * -KA * "].concat(BOARD3_9),
        turn: Color.White,
    },
    KY_R: {
        board: [" * -KE-GI-KI-OU-KI-GI-KE-KY", " * -HI *  *  *  *  * -KA * "].concat(BOARD3_9),
        turn: Color.White,
    },
    KA: {
        board: ["-KY-KE-GI-KI-OU-KI-GI-KE-KY", " * -HI *  *  *  *  *  *  * "].concat(BOARD3_9),
        turn: Color.White,
    },
    HI: {
        board: ["-KY-KE-GI-KI-OU-KI-GI-KE-KY", " *  *  *  *  *  *  * -KA * "].concat(BOARD3_9),
        turn: Color.White,
    },
    HIKY: {
        board: ["-KY-KE-GI-KI-OU-KI-GI-KE * ", " *  *  *  *  *  *  * -KA * "].concat(BOARD3_9),
        turn: Color.White,
    },
    "2": {
        board: ["-KY-KE-GI-KI-OU-KI-GI-KE-KY"].concat(BOARD2_9),
        turn: Color.White,
    },
    "3": {
        board: ["-KY-KE-GI-KI-OU-KI-GI-KE * "].concat(BOARD2_9),
        turn: Color.White,
    },
    "4": {
        board: [" * -KE-GI-KI-OU-KI-GI-KE * "].concat(BOARD2_9),
        turn: Color.White,
    },
    "5": {
        board: [" *  * -GI-KI-OU-KI-GI-KE * "].concat(BOARD2_9),
        turn: Color.White,
    },
    "5_L": {
        board: [" * -KE-GI-KI-OU-KI-GI *  * "].concat(BOARD2_9),
        turn: Color.White,
    },
    "6": {
        board: [" *  * -GI-KI-OU-KI-GI *  * "].concat(BOARD2_9),
        turn: Color.White,
    },
    "7_R": {
        board: [" *  *  * -KI-OU-KI-GI *  * "].concat(BOARD2_9),
        turn: Color.White,
    },
    "7_L": {
        board: [" *  * -GI-KI-OU-KI *  *  * "].concat(BOARD2_9),
        turn: Color.White,
    },
    "8": {
        board: [" *  *  * -KI-OU-KI *  *  * "].concat(BOARD2_9),
        turn: Color.White,
    },
    "10": {
        board: [" *  *  *  * -OU *  *  *  * "].concat(BOARD2_9),
        turn: Color.White,
    },
};

export type IPreset = keyof typeof presetDefinitions;
export const getInitialFromPreset = (preset: string) => {
    const definition = presetDefinitions[preset];
    if (!definition) {
        throw new Error(`Unknown preset: ${preset}`);
    }
    return definition;
};

const presets = Object.keys(presetDefinitions);
export default presets;
