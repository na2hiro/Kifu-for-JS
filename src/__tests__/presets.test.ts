import presets, {getInitialFromPreset} from "../presets";

describe("preset", () => {
    it("initialized board", () => {
        for (const preset of presets) {
            expect(getInitialFromPreset(preset)).toMatchSnapshot(preset);
        }
    });
});
