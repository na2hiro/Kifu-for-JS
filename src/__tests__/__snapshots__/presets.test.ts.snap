// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preset initialized board: 2 1`] = `
Object {
  "board": Array [
    "-KY-KE-GI-KI-OU-KI-GI-KE-KY",
    " *  *  *  *  *  *  *  *  * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: 3 1`] = `
Object {
  "board": Array [
    "-KY-KE-GI-KI-OU-KI-GI-KE * ",
    " *  *  *  *  *  *  *  *  * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: 4 1`] = `
Object {
  "board": Array [
    " * -KE-GI-KI-OU-KI-GI-KE * ",
    " *  *  *  *  *  *  *  *  * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: 5 1`] = `
Object {
  "board": Array [
    " *  * -GI-KI-OU-KI-GI-KE * ",
    " *  *  *  *  *  *  *  *  * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: 5_L 1`] = `
Object {
  "board": Array [
    " * -KE-GI-KI-OU-KI-GI *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: 6 1`] = `
Object {
  "board": Array [
    " *  * -GI-KI-OU-KI-GI *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: 7_L 1`] = `
Object {
  "board": Array [
    " *  * -GI-KI-OU-KI *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: 7_R 1`] = `
Object {
  "board": Array [
    " *  *  * -KI-OU-KI-GI *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: 8 1`] = `
Object {
  "board": Array [
    " *  *  * -KI-OU-KI *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: 10 1`] = `
Object {
  "board": Array [
    " *  *  *  * -OU *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: HI 1`] = `
Object {
  "board": Array [
    "-KY-KE-GI-KI-OU-KI-GI-KE-KY",
    " *  *  *  *  *  *  * -KA * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: HIKY 1`] = `
Object {
  "board": Array [
    "-KY-KE-GI-KI-OU-KI-GI-KE * ",
    " *  *  *  *  *  *  * -KA * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: HIRATE 1`] = `
Object {
  "board": Array [
    "-KY-KE-GI-KI-OU-KI-GI-KE-KY",
    " * -HI *  *  *  *  * -KA * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 0,
}
`;

exports[`preset initialized board: KA 1`] = `
Object {
  "board": Array [
    "-KY-KE-GI-KI-OU-KI-GI-KE-KY",
    " * -HI *  *  *  *  *  *  * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: KY 1`] = `
Object {
  "board": Array [
    "-KY-KE-GI-KI-OU-KI-GI-KE * ",
    " * -HI *  *  *  *  * -KA * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;

exports[`preset initialized board: KY_R 1`] = `
Object {
  "board": Array [
    " * -KE-GI-KI-OU-KI-GI-KE-KY",
    " * -HI *  *  *  *  * -KA * ",
    "-FU-FU-FU-FU-FU-FU-FU-FU-FU",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    " *  *  *  *  *  *  *  *  * ",
    "+FU+FU+FU+FU+FU+FU+FU+FU+FU",
    " * +KA *  *  *  *  * +HI * ",
    "+KY+KE+GI+KI+OU+KI+GI+KE+KY",
  ],
  "turn": 1,
}
`;
